transports/quic/src/endpoint.rs

```diff
@@ -403,17 +403,28 @@ where
         if me.event_slot.is_none() && me.incoming_slot.is_none() {
             let mut metas = [RecvMeta::default(); BATCH_SIZE];
             let mut iovs = MaybeUninit::<[IoSliceMut; BATCH_SIZE]>::uninit();
-            me.recv_buf
-                .chunks_mut(me.recv_buf.len() / BATCH_SIZE)
-                .enumerate()
-                .for_each(|(i, buf)| unsafe {
+            // SAFETY: safe as long as iovs do not outlive recv_buf
+            // and there exist only one iovs initialized with the given recv_buf.
+            // Also recv_buf should not be moved while there exist initialized iovs.
```

Comment:
Is there a reason Rust cannot track these constraints at compile-time?

Reply:
We can use 'a everywhere but the borrowchecker won't allow us to call to any &mut self method as long as &mut iovs lives.

Reply:
Why do such methods need to be called?  This seems like it can be solved with some refactoring.

Reply:
```
error[E0499]: cannot borrow `*me` as mutable more than once at a time
   --> transports/quic/src/endpoint.rs:448:32
    |
426 |             let iovs = unsafe { init_iovs(&mut iovs, &mut me.recv_buf) };
    |                                                      ---------------- first mutable borrow occurs here
427 |             while let Poll::Ready(result) = me.socket.poll_recv(cx, iovs, &mut metas) {
    |                                                                     ---- first borrow later used here
...
448 |                             if me.send_event(id, event, cx) {
    |                                ^^ second mutable borrow occurs here

```

Reply:
Can this be at least tested in Miri?  Right now this looks very suspicious.  Would it be possible for `me.send_event` to be refactored so as to only take the relevant parts of `me`?

Reply:
It worked before because `iov` was 'unbounded to `self` lifetime. Imagine:

```
let mut iovs = MaybeUninit::<[IoSliceMut<'unbounded>; BATCH_SIZE]>::uninit();
```

I can't make an unbounded lifetime in args due to lifetime elision. To either we trick rustc with `iovs` being bounded to itself or trick with an unbounded lifetime without factoring the code in a separate function.

Reply:
> as to only take the relevant parts of me?

Unfortunately no as long as recv_buf is a part of me/self. rustc borrows all fields in methods.

Reply:
Well there are multiple solutions:
use mem::replace to take the buffer out of the struct or turn send_event into a function and restructure self into mutable references to it's fields

Reply:
Yes... But do we need all this?

Reply:
Well we've gotten this far, so may as well finish the job. Personally I was fine with how it was ...

Reply:
> use mem::replace to take the buffer out of the struct or turn send_event into a function and destructure self into mutable references to it's fields

That was what I was thinking of, yes.